File: src/ladym/operations/l6.py
"""L6 forward-intent prediction (SPEC §2.8).

From recent L1 episodes, predict likely next intents via the ``l6_forward_intent`` agent and
store one ``L6_predictive`` memory per intent. Each prediction carries ``metadata.valid_to``;
at the start of every run, expired predictions are retired (DELETE-style, no successor) so
stale intents drop out of recall but stay auditable.
"""
from __future__ import annotations

import time
from dataclasses import dataclass, field

from pydantic import BaseModel

from ..config import Config
from ..schema import Layer, Memory, MemoryType
from ..storage.embeddings import EmbeddingProvider
from ..storage.store import SQLiteStore
from .supersedes import is_retired, retire

_L6_LAYER = Layer.L6_PREDICTIVE.value
_L1_LAYER = Layer.EPISODIC.value
_WATERMARK = "l6_last_episode_ts"


class _Intent(BaseModel):
    intent: str
    confidence: float = 0.5
    horizon_s: float | None = None


class _Intents(BaseModel):
    intents: list[_Intent]


@dataclass
class L6PredictionReport:
    predictions: int = 0
    expired_retired: int = 0
    episodes_seen: int = 0
    watermark_updated_to: float | None = None
    details: list[dict] = field(default_factory=list)
    skipped: bool = False


def _default_prompt() -> str:
    from importlib.resources import files

    return (files("ladym.prompts") / "l6.txt").read_text()


def predict(
    store: SQLiteStore,
    embedder: EmbeddingProvider,
    *,
    cfg: Config,
    workspace: str | None = None,
    llm=None,
    prompt: str | None = None,
) -> L6PredictionReport:
    ws = workspace or cfg.workspace
    report = L6PredictionReport()
    if llm is None:
        report.skipped = True
        return report
    prompt = prompt or _default_prompt()
    now = time.time()

    # 1. expire sweep: retire predictions whose valid_to has passed.
    for m in store.iter_memories(workspace=ws, layer=_L6_LAYER):
        if is_retired(m):
            continue
        try:
            valid_to = float(m.metadata.get("valid_to", 0))
        except (TypeError, ValueError):
            continue
        if now > valid_to:
            retire(store, m)
            report.expired_retired += 1

    # 2. window: episodes newer than the watermark, capped.
    watermark = float(store.get_meta(_WATERMARK) or 0.0)
    rows = store.conn.execute(
        "SELECT * FROM memories WHERE layer = ? AND workspace = ? AND created_at > ? "
        "ORDER BY created_at ASC LIMIT ?",
        (_L1_LAYER, ws, watermark, cfg.system2.l6_max_episodes),
    ).fetchall()
    episodes = [store._row_to_memory(r) for r in rows]
    report.episodes_seen = len(episodes)
    if not episodes:
        return report

    # 3. predict.
    corpus = "\n".join(f"- {m.content}" for m in episodes)
    user_msg = (
        "Predict likely next intents from these recent episodes:\n" + corpus
    )
    msgs = [
        {"role": "system", "content": prompt},
        {"role": "user", "content": user_msg},
    ]
    out = llm.complete_structured(msgs, _Intents)
    out = out if isinstance(out, dict) else out.model_dump()
    default_horizon = cfg.system2.l6_horizon_s

    # 4. store one memory per intent.
    for it in out.get("intents", []):
        intent_text = (it.get("intent") or "").strip() if isinstance(it, dict) else ""
        if not intent_text:
            continue
        confidence = float(it.get("confidence", 0.5))
        horizon = it.get("horizon_s")
        horizon = default_horizon if horizon is None else float(horizon)
        valid_to = now + horizon
        mem = Memory(
            layer=Layer.L6_PREDICTIVE,
            type=MemoryType.FORWARD_INTENT,
            content=intent_text,
            summary=intent_text[:80],
            tags=["predicted"],
            metadata={"confidence": confidence, "horizon_s": horizon, "valid_to": valid_to},
            source="l6_predict",
            workspace=ws,
        )
        store.put_memory(mem, vector=embedder.embed(intent_text))
        report.predictions += 1
        report.details.append(
            {"intent": intent_text, "confidence": confidence, "valid_to": valid_to}
        )

    # 5. advance the watermark past the batch we just consumed.
    newest = max(m.created_at for m in episodes)
    store.set_meta(_WATERMARK, str(newest))
    report.watermark_updated_to = newest
    return report

File: operations/l6.go
package operations

import (
	"sort"
	"strings"

	"github.com/ProjAnvil/LadyM/config"
	"github.com/ProjAnvil/LadyM/providers"
	"github.com/ProjAnvil/LadyM/schema"
	"github.com/ProjAnvil/LadyM/storage"
)

const l6Layer = schema.LayerL6Predictive
const l6Watermark = "l6_last_episode_ts"

const l6DefaultPrompt = `You are a forward-prediction engine for a brain-inspired agent memory system. You are given a
list of recent episodic events (what the agent just did or observed). Predict the most likely
NEXT intents the agent or user will pursue.

Reply ONLY with JSON matching exactly this schema:
  {"intents": [{"intent": "<a concrete next action or goal>", "confidence": <0.0-1.0>, "horizon_s": <seconds this prediction stays plausible>}]}

Rules:
- Return between 1 and 5 intents, most likely first.
- "intent" must be a concrete next step, not a vague topic.
- You may omit "horizon_s" on an intent only if you cannot estimate it; the system defaults it.
`

// L6PredictionReport reports the outcome of L6 forward-intent prediction.
type L6PredictionReport struct {
	Predictions        int
	ExpiredRetired     int
	EpisodesSeen       int
	WatermarkUpdatedTo float64
	Details            []map[string]any
	Skipped            bool
}

// PredictL6 predicts next intents from recent episodes, with TTL expiry.
func PredictL6(store storage.Store, embedder storage.EmbeddingProvider, cfg *config.Config, workspace string, llm providers.LLMProvider, prompt string) (*L6PredictionReport, error) {
	ws := workspace
	if ws == "" {
		ws = cfg.Workspace
	}
	report := &L6PredictionReport{}
	if llm == nil {
		report.Skipped = true
		return report, nil
	}
	if prompt == "" {
		prompt = l6DefaultPrompt
	}
	now := schema.Now()

	// 1. expire sweep
	l6s, err := store.IterMemories(ws, string(l6Layer), "")
	if err != nil {
		return nil, err
	}
	for _, m := range l6s {
		if IsRetired(m) {
			continue
		}
		// Python: float(m.metadata.get("valid_to", 0)) — missing counts as 0
		// (expired); a present but non-numeric value raises and is skipped.
		validTo, ok := m.MetaFloat("valid_to")
		if !ok {
			if _, exists := m.Metadata["valid_to"]; exists {
				continue
			}
			validTo = 0
		}
		if now > validTo {
			if err := Retire(store, m, ""); err != nil {
				return nil, err
			}
			report.ExpiredRetired++
		}
	}

	// 2. window
	rawWatermark, err := store.GetMeta(l6Watermark)
	if err != nil {
		return nil, err
	}
	watermark := 0.0
	if rawWatermark != "" {
		watermark = parseFloatOrZero(rawWatermark)
	}
	episodes, err := store.IterMemories(ws, string(schema.LayerEpisodic), "")
	if err != nil {
		return nil, err
	}
	sort.SliceStable(episodes, func(a, b int) bool { return episodes[a].CreatedAt < episodes[b].CreatedAt })
	var recent []*schema.Memory
	for _, e := range episodes {
		if e.CreatedAt > watermark {
			recent = append(recent, e)
		}
	}
	if len(recent) > cfg.System2.L6MaxEpisodes {
		recent = recent[:cfg.System2.L6MaxEpisodes]
	}
	report.EpisodesSeen = len(recent)
	if len(recent) == 0 {
		return report, nil
	}

	// 3. predict
	lines := make([]string, 0, len(recent))
	for _, e := range recent {
		lines = append(lines, "- "+e.Content)
	}
	corpus := strings.Join(lines, "\n")
	msgs := []providers.Message{
		{Role: "system", Content: prompt},
		{Role: "user", Content: "Predict likely next intents from these recent episodes:\n" + corpus},
	}
	out, err := llm.CompleteStructured(msgs, providers.JSONSchema{
		"type": "object",
		"properties": map[string]any{
			"intents": map[string]any{
				"type":        "array",
				"description": "Predicted next intents",
				"items": map[string]any{
					"type": "object",
					"properties": map[string]any{
						"intent":     map[string]any{"type": "string", "description": "Predicted intent text"},
						"confidence": map[string]any{"type": "number", "description": "0..1 confidence, default 0.5"},
						"horizon_s":  map[string]any{"type": "number", "description": "Seconds the prediction stays valid; omit for the configured default"},
					},
					"required": []string{"intent"},
				},
			},
		},
		"required": []string{"intents"},
	})
	if err != nil {
		return nil, err
	}
	defaultHorizon := cfg.System2.L6HorizonS

	intents, _ := out["intents"].([]any)
	for _, it := range intents {
		obj, ok := it.(map[string]any)
		if !ok {
			continue
		}
		intentText, _ := obj["intent"].(string)
		intentText = strings.TrimSpace(intentText)
		if intentText == "" {
			continue
		}
		confidence := 0.5
		if c, ok := obj["confidence"].(float64); ok {
			confidence = c
		}
		horizon := defaultHorizon
		if h, ok := obj["horizon_s"].(float64); ok {
			horizon = h
		}
		validTo := now + horizon
		m := schema.NewMemory(l6Layer, schema.TypeForwardIntent)
		m.Content = intentText
		m.Summary = truncateStr(intentText, 80)
		m.Tags = []string{"predicted"}
		m.Metadata = map[string]any{"confidence": confidence, "horizon_s": horizon, "valid_to": validTo}
		m.Source = "l6_predict"
		m.Workspace = ws
		vec, err := embedder.Embed(intentText)
		if err != nil {
			return nil, err
		}
		if err := store.PutMemory(m, vec); err != nil {
			return nil, err
		}
		report.Predictions++
		report.Details = append(report.Details, map[string]any{
			"intent": intentText, "confidence": confidence, "valid_to": validTo,
		})
	}

	newest := recent[len(recent)-1].CreatedAt
	if err := store.SetMeta(l6Watermark, floatString(newest)); err != nil {
		return nil, err
	}
	report.WatermarkUpdatedTo = newest
	return report, nil
}

func truncateStr(s string, n int) string {
	r := []rune(s)
	if len(r) <= n {
		return s
	}
	return string(r[:n])
}
